src/init.cpp

```diff
@@ -274,6 +280,18 @@ void Shutdown()
         DumpMempool();
     }
 
+    if (gArgs.GetBoolArg("-watchonly", false)) {
+        if (!FlushWatchOnlyAddresses()) {
+            error("Fail to write watchonly addresses to database.");
+        }
+
+        {
+            LOCK(cs_watchonly);
+            // Ran into memory problems if this map wasn't cleared before shutdown.
```

Comment:
That sounds like there may be another issue, perhaps this is a global? This would be better encapsulated somewhere in a destructor if not.